server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java

```diff
@@ -300,12 +301,20 @@ public String createId(Map<String, Object> flat, byte[] suffix) {
             Builder b = builder();
             for (Map.Entry<String, Object> e : flat.entrySet()) {
                 if (isRoutingPath.test(e.getKey())) {
-                    b.hashes.add(new NameAndHash(new BytesRef(e.getKey()), hash(new BytesRef(e.getValue().toString()))));
+                    if (e.getValue() instanceof List<?> listValue) {
+                        listValue.forEach(v -> hashValue(b, e.getKey(), v));
```

Comment:
Nit: stream calls are slightly inefficient, iiuc? If so, replace with a regular for-loop.